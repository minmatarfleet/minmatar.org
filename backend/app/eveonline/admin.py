from django.contrib import admin

from .models import (
    EveAlliance,
    EveCharacter,
    EveCharacterSkillset,
    EveCorporation,
    EveCorporationApplication,
)

# Register your models here.
admin.site.register(EveCorporation)
admin.site.register(EveCharacter)
admin.site.register(EveCharacterSkillset)
admin.site.register(EveCorporationApplication)
admin.site.register(EveAlliance)
