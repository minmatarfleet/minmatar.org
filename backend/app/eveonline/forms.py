from crispy_forms.helper import FormHelper
from crispy_forms.layout import Submit, Layout
from .models import EveCorporationApplication
from django.forms import ModelForm


class EveCorporationApplicationForm(ModelForm):
    class Meta:
        model = EveCorporationApplication
        fields = [
            "description",
        ]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.helper = FormHelper()
        self.helper.add_input(Submit("submit", "Submit"))
