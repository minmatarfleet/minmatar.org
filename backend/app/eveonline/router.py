from typing import List

from authentication import AuthBearer
from ninja import ModelSchema, Router, Schema

from .models import EveCorporation

router = Router(tags=["Eve Online"])


class CorporationResponse(ModelSchema):
    class Meta:
        model = EveCorporation
        fields = "__all__"


class CreateCorporationRequest(Schema):
    corporation_id: int


@router.get(
    "/corporations", response=List[CorporationResponse], auth=AuthBearer()
)
def get_corporations(request):  # pylint: disable=unused-argument
    return EveCorporation.objects.all()


@router.get(
    "/corporations/{corporation_id}",
    response=CorporationResponse,
    auth=AuthBearer(),
)
def get_corporation_by_id(
    request, corporation_id: int
):  # pylint: disable=unused-argument
    return EveCorporation.objects.get(corporation_id=corporation_id)


@router.post("/corporations", response=CorporationResponse, auth=AuthBearer())
def create_corporation(
    request, payload: CreateCorporationRequest
):  # pylint: disable=unused-argument
    return EveCorporation.objects.create(**payload.dict())
