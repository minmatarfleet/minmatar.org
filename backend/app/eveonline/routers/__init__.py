from eveonline.routers.corporations import router as corporations_router
from eveonline.routers.characters import router as characters_router
from ninja import Router

router = Router(tags=["Eve Online"])
router.add_router("corporations/", corporations_router)
router.add_router("characters/", characters_router)
