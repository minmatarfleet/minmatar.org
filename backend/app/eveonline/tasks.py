import logging

from esi.clients import EsiClientProvider
from esi.models import Token

from app.celery import app

from .models import EveCharacter, EveCorporation

esi = EsiClientProvider()
logger = logging.getLogger(__name__)


@app.task
def update_corporations():
    for corporation in EveCorporation.objects.all():
        logger.info("Updating corporation %s", corporation.name)
        corporation.save()
        if not corporation.active():
            logger.info(
                "Skipping extra steps for inactive corporation %s",
                corporation.name,
            )
            continue

        logger.info("Updating corporation %s characters", corporation.name)
        for character in EveCharacter.objects.filter(
            corporation_id=corporation.corporation_id
        ):
            logger.info("Updating character %s", character.character_name)
            character.save()

        logger.info(
            "Updating corporation %s from external roster", corporation.name
        )
        required_scopes = ["esi-corporations.read_corporation_membership.v1"]
        token = Token.objects.get(
            character_id=corporation.ceo_id, scopes__name__in=required_scopes
        )
        esi_corporation_members = (
            esi.client.Corporation.get_corporations_corporation_id_members(
                corporation_id=corporation.corporation_id,
                token=token.valid_access_token(),
            ).results()
        )

        logger.info("Found %s members", len(esi_corporation_members))
        for character_id in esi_corporation_members:
            if not EveCharacter.objects.filter(
                character_id=character_id
            ).exists():
                logger.info("Creating character %s", character_id)
                character = EveCharacter.objects.create(
                    character_id=character_id
                )
