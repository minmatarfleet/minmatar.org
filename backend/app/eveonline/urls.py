"""
URL configuration for app project.

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/5.0/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.urls import path
from . import views

urlpatterns = [
    path("characters", views.list_characters, name="eveonline-characters"),
    path(
        "characters/<int:character_id>/view",
        views.view_character,
        name="eveonline-characters-view",
    ),
    path(
        "characters/add", views.add_character, name="eveonline-characters-add"
    ),
    path(
        "characters/add-alliance-character",
        views.add_alliance_character,
        name="eveonline-characters-add-alliance-character",
    ),
    path(
        "corporations", views.list_corporations, name="eveonline-corporations"
    ),
    path(
        "corporations/<int:corporation_pk>/view",
        views.view_corporation,
        name="eveonline-corporations-view",
    ),
    path(
        "corporations/<int:corporation_pk>/update",
        views.update_corporation,
        name="eveonline-corporations-update",
    ),
    path(
        "corporations/<int:corporation_pk>/apply",
        views.create_corporation_application,
        name="eveonline-corporations-apply",
    ),
    path(
        "corporations/applications/<int:application_pk>",
        views.view_corporation_application,
        name="eveonline-corporations-applications-view",
    ),
    path(
        "corporations/applications/<int:application_pk>/approve",
        views.approve_corporation_application,
        name="eveonline-corporations-applications-approve",
    ),
    path(
        "corporations/applications/<int:application_pk>/reject",
        views.reject_corporation_application,
        name="eveonline-corporations-applications-reject",
    ),
]
