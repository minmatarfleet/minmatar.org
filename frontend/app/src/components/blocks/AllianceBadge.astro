---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { AllianceBadgeProps } from '@dtypes/layout_components'
import { get_alliance_by_id } from '@helpers/api.eveonline/alliances'
import type { AllianceEvE } from '@dtypes/api.eveonline'
import { prod_error_messages } from '@helpers/env'

interface Props {
    alliance: AllianceBadgeProps;
}

const {
    alliance,
} = Astro.props

let alliance_eve:AllianceEvE | null = null
let get_eve_alliance_error
if (!alliance?.name) {
    try {
        alliance_eve = await get_alliance_by_id(alliance.id)
    } catch (error) {
        get_eve_alliance_error = (prod_error_messages() ? t('get_eve_alliance_error') : error.message)
    }
}

const alliance_name = alliance?.name ?? (alliance_eve?.name ?? get_eve_alliance_error)

import { get_alliance_logo } from '@helpers/eve_image_server';
import FixedFluid from '@components/compositions/FixedFluid.astro'
---
<FixedFluid
    width='32px'
    gap='var(--space-2xs)'
    class="[ items-center ]"
>
    <picture>
        <img loading="lazy" src={get_alliance_logo(alliance.id)} width="32" height="32" />
    </picture>
    <small>
        <span class="[ text-[var(--highlight)] ]">{alliance_name}</span>
    </small>
</FixedFluid>