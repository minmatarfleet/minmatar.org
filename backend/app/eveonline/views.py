from django.shortcuts import render, redirect
from django.contrib.auth.decorators import login_required, permission_required
from django.contrib.auth.models import User
from esi.decorators import token_required
from eveonline.helpers.skills import compare_skills_to_skillset
from .models import (
    EvePrimaryToken,
    EveCorporation,
    EveCorporationApplication,
    EveCharacter,
    EveCharacterSkillset,
)
from .forms import EveCorporationApplicationForm
from pydantic import BaseModel
from typing import List
from django.contrib import messages
from eveonline.helpers.characters import (
    get_character_list,
    get_corporation_character_list,
    ALLIANCE_SCOPES,
    TokenType,
)
from datetime import timedelta
from django.utils import timezone


class CharacterResponse(BaseModel):
    character_id: int
    character_name: str
    type: str = "Public"
    scopes: List[str]
    is_primary: bool = False


class CorporationResponse(BaseModel):
    alliance_section: List[List[dict]]
    associate_section: List[List[dict]]
    militia_section: List[List[dict]]


# Create your views here.
@login_required
def list_characters(request):
    characters = get_character_list(request.user)
    context = {
        "characters": characters,
    }

    return render(
        request, "eveonline/characters/list_characters.html", context
    )


@login_required
@permission_required("eveonline.view_evecharacter")
def view_character(request, character_id):
    character = EveCharacter.objects.get(character_id=character_id)
    skillsets = EveCharacterSkillset.objects.all()

    skillsets_response = []
    for skillset in skillsets:
        missing_skills, progress = compare_skills_to_skillset(
            character, skillset
        )
        progress_color = "success"
        if progress < 50:
            progress_color = "danger"
        elif progress < 100:
            progress_color = "warning"
        skillsets_response.append(
            {
                "skillset": skillset,
                "missing_skills": "\n".join(missing_skills),
                "progress": {"value": progress, "color": progress_color},
            }
        )

    context = {
        "character": character,
        "skillsets": skillsets_response,
    }

    return render(request, "eveonline/characters/view_character.html", context)


@login_required
@token_required()
def add_character(request, token):
    user = User.objects.get(pk=request.user.id)
    if not EvePrimaryToken.objects.filter(user=user).exists():
        EvePrimaryToken.objects.create(user=user, token=token)
    return redirect("eveonline-characters")


@login_required
@token_required(scopes=ALLIANCE_SCOPES)
def add_alliance_character(request, token):
    next = request.GET.get("next", None)
    user = User.objects.get(pk=request.user.id)
    if not EvePrimaryToken.objects.filter(user=user).exists():
        EvePrimaryToken.objects.create(user=user, token=token)
    if next:
        return redirect(next)
    return redirect("eveonline-characters")


def list_corporations(request):
    n = 3
    alliance_corporations = EveCorporation.objects.filter(type="alliance")
    # break into n-sized chunks
    alliance_corporations = [
        alliance_corporations[i : i + n]
        for i in range(0, len(alliance_corporations), n)
    ]

    associate_corporations = EveCorporation.objects.filter(type="associate")
    # break into n-sized chunks
    associate_corporations = [
        associate_corporations[i : i + n]
        for i in range(0, len(associate_corporations), n)
    ]

    militia_corporations = EveCorporation.objects.filter(type="militia")
    # break into n-sized chunks
    militia_corporations = [
        militia_corporations[i : i + n]
        for i in range(0, len(militia_corporations), n)
    ]

    context = {
        "alliance_corporation_chunks": alliance_corporations,
        "associate_corporation_chunks": associate_corporations,
        "militia_corporation_chunks": militia_corporations,
    }

    return render(
        request, "eveonline/corporations/list_corporations.html", context
    )


def view_corporation(request, corporation_pk):
    corporation = EveCorporation.objects.get(pk=corporation_pk)
    characters = get_corporation_character_list(corporation)
    context = {
        "corporation": corporation,
        "characters": characters,
    }
    return render(
        request, "eveonline/corporations/view_corporation.html", context
    )


@login_required
@token_required(scopes=EveCorporation.required_ceo_scopes)
def update_corporation(request, token, corporation_pk):
    corporation = EveCorporation.objects.get(pk=corporation_pk)
    if token.character_id == corporation.ceo_id:
        messages.success(
            request, "Successfully added CEO token for corporation"
        )
    else:
        messages.error(request, "Must be the CEO to add token for corporation")
    return redirect("eveonline-corporations")


@login_required
def create_corporation_application(request, corporation_pk):
    corporation = EveCorporation.objects.get(pk=corporation_pk)
    characters = get_character_list(request.user)
    if EveCorporationApplication.objects.filter(
        corporation=corporation,
        user=request.user,
        created_at__gt=timezone.now() - timedelta(days=30),
    ).exists():
        existing_application = EveCorporationApplication.objects.get(
            corporation=corporation,
            user=request.user,
        )
    else:
        existing_application = None
    for character in characters:
        if character.type != TokenType.ALLIANCE:
            messages.warning(
                request,
                f"Token for {character.character_name} has an improper token type. Please add this character again",
            )
    if request.method == "POST":
        submit = True
        form = EveCorporationApplicationForm(request.POST)
        # Check characters
        if not characters:
            messages.error(
                request,
                "You must have at least one character to apply to a corporation",
            )
            submit = False

        for character in characters:
            if character.type != TokenType.ALLIANCE:
                submit = False

        # Check form
        if not form.is_valid():
            submit = False

        if submit:
            application = form.save(commit=False)
            application.corporation = EveCorporation.objects.get(
                pk=corporation_pk
            )
            application.user = request.user
            application.save()
            messages.info(request, "Successfully applied to corporation")
            return redirect("eveonline-corporations-apply", corporation_pk)
    else:
        form = EveCorporationApplicationForm()

    context = {
        "form": form,
        "characters": characters,
        "corporation": corporation,
        "existing_application": existing_application,
    }
    return render(
        request,
        "eveonline/applications/create_corporation_application.html",
        context,
    )


@login_required
@permission_required("eveonline.view_evecorporationapplication")
def view_corporation_application(request, application_pk):
    application = EveCorporationApplication.objects.get(pk=application_pk)
    discord_user = request.user.discord_user
    characters = get_character_list(application.user)
    context = {
        "application": application,
        "discord_user": discord_user,
        "characters": characters,
        "discord_thread_link": f"https://discord.com/channels/1041384161505722368/{application.discord_thread_id}",
    }
    return render(
        request,
        "eveonline/applications/view_corporation_application.html",
        context,
    )


@login_required
@permission_required("eveonline.view_evecorporationapplication")
def approve_corporation_application(request, application_pk):
    application = EveCorporationApplication.objects.get(pk=application_pk)
    application.status = "accepted"
    application.save()
    return redirect("eveonline-corporations-applications-view", application_pk)


@login_required
@permission_required("eveonline.view_evecorporationapplication")
def reject_corporation_application(request, application_pk):
    application = EveCorporationApplication.objects.get(pk=application_pk)
    application.status = "rejected"
    application.save()
    return redirect("eveonline-corporations-applications-view", application_pk)
