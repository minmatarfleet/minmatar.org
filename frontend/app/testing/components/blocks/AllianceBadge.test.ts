import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi } from "vitest";

import AllianceBadge from "@components/blocks/AllianceBadge.astro";

import { get_alliance_by_id } from "@helpers/api.eveonline/alliances";
import { get_alliance_logo } from "@helpers/eve_image_server";

vi.mock("@helpers/api.eveonline/alliances");
vi.mock("@helpers/eve_image_server");

test("AllianceBadge defaults", async () => {
  vi.mocked(get_alliance_by_id).mockImplementation(async (id) => {
    expect(id).toBe(42);

    return {
      name: "lolcat alliance is best alliance",
      creator_corporation_id: 1,
      creator_id: 1,
      date_founded: new Date(),
      executor_corporation_id: 1,
      faction_id: 1,
      ticker: "TICK",
    };
  });

  vi.mocked(get_alliance_logo).mockImplementation((id, size) => {
    expect(id).toBe(42);
    expect(size).toBe(32);

    return "test-alliance-image-src";
  });

  const container = await AstroContainer.create();
  const result = await container.renderToString(AllianceBadge, {
    props: {
      alliance: {
        id: 42,
      },
    },
  });

  expect(result).toMatchSnapshot();
});
