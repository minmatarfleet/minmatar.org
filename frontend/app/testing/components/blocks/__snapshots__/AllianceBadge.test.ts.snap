// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AllianceBadge defaults 1`] = `"<div class="fixed-fluid [ items-center ]" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-2xs);--fixed_fluid_width: smpx;--fixed_fluid_breakpoint: 50%;">  <picture> <img loading="lazy" src="test-alliance-image-src" width="32" height="32"> </picture> <small class="[ w-full ]"> <span class="[ text-[var(--highlight)] ]">lolcat alliance is best alliance</span> </small>  </div>  "`;
